finality-aleph/src/justification/scheduler.rs

```diff
@@ -0,0 +1,85 @@
+use crate::{MillisecsPerBlock, SessionPeriod};
+use std::{
+    cmp::min,
+    time::{Duration, Instant},
+};
+
+pub enum SchedulerActions {
+    ClearQueue,
+    Request,
+    Wait,
+}
+
+/// Bunch of methods for managing frequency of sending justification requests.
+pub trait JustificationRequestScheduler {
+    /// Decides whether we can request new justification.
+    fn schedule_action(&mut self) -> SchedulerActions;
+    /// Notice block finalization.
+    fn on_block_finalized(&mut self);
+    /// Notice request sending.
+    fn on_request_sent(&mut self);
+}
+
+pub struct JustificationRequestSchedulerImpl {
+    last_request_time: Instant,
+    last_finalization_time: Instant,
+    delay: Duration,
+    attempt: u32,
+    max_attemps: u32,
+}
+
+impl JustificationRequestSchedulerImpl {
+    pub fn new(
+        session_period: &SessionPeriod,
+        millisecs_per_block: &MillisecsPerBlock,
+        max_attemps: u32,
+    ) -> Self {
+        Self {
+            last_request_time: Instant::now(),
+            last_finalization_time: Instant::now(),
+            delay: Duration::from_millis(min(
+                millisecs_per_block.0 * 2,
+                millisecs_per_block.0 * session_period.0 as u64 / 10,
+            )),
+            attempt: 0,
+            max_attemps,
+        }
+    }
+}
+
+impl JustificationRequestSchedulerImpl {
```

Comment:
nit: this could be in the `impl` block above?

Reply:
Done